src/replicache.ts

```diff
@@ -31,16 +31,36 @@ export const httpStatusUnauthorized = 401;
 
 export type MaybePromise<T> = T | Promise<T>;
 
+type ToPromise<P> = P extends Promise<unknown> ? P : Promise<P>;
+
 /** The key name to use in localStorage when synchronizing changes. */
 const storageKeyName = (name: string) => `/replicache/root/${name}`;
 
 /** The maximum number of time to call out to getDataLayerAuth before giving up and throwing an error. */
 const MAX_REAUTH_TRIES = 8;
 
+/**
+ * The type used when implementing mutators.
+ */
+export type MutatorImpl<
+  // Not sure how to not use any here...
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  Args extends JSONValue = any,
+  Return extends JSONValue | void = JSONValue | void
+> = (tx: WriteTransaction, args?: Args) => MaybePromise<Return>;
+
+/**
+ * The type used to describe the mutator definitions passed into [[Replicache]]
+ * constructor as part of the [[ReplicacheOptions]].
+ */
+export type MutatorDefs = {
+  [key: string]: MutatorImpl;
+};
+
 /**
  * The options passed to [[Replicache]].
  */
-export interface ReplicacheOptions {
+export interface ReplicacheOptions<MD extends MutatorDefs> {
```

Comment:
This is amazing, I had no idea you could _extend_ a dictionary in this way. So this type is both dictionary-y and we can access as key/value pairs, and as a strongly typed interface. So cool.

Do we need to worry about the typescript compiler shortening the field names? Because like dev only ever access them as `foo.bar`, maybe TSC won't know that the "bar" should be preserved literally?

I feel like that's something you need to worry about with closure compiler.

Reply:
### Extends

Not sure I understand what you understand here? There is no distinction between a dictionary and an interface. TS is all structural. The only time this becomes hazy is when you have private properties in a class. The syntax `{[key: string]: V}` means that all the keys must be a string (and not a symbol or a number) and all the `Value`s must be of type `V`. When you say that the type parameter `extends` something it means that the concrete type must be a subtype of that. 

The problem then becomes what happens if I say I want a function that takes an argument that is `JSONValue`. Well if the concrete type says it wants a `number` then that is not safe because we said that we might call it with a `JSONValue` :'(

Some languages have ways to force covariance and contravariance but TS tries to be smart and pick the right one.

### Renaming

tsc does not rename property names. esbuild does not rename property names. terser does not rename property names.

Only Closure Compiler renames property names but only in advanced mode and advanced mode is not compatible with TypeScript's type system.

Reply:
thanks for explaining

Reply:
> thanks for explaining

I tried